File: src/test/scala/workshop/part2/ClientActorTest.scala
package workshop.part2

import akka.actor.{Terminated, _}
import akka.testkit.{TestActorRef, TestProbe}
import workshop.AkkaSpec
import workshop.helpers.ComputeTestActor
import workshop.work.{RiskyAddition, RiskyAdditionResult, RiskyWork, RiskyWorkException}

import scala.concurrent.duration.Duration.Zero
import scala.concurrent.duration._
import scala.language.postfixOps

class ClientActorTest extends AkkaSpec {

  val timeout: FiniteDuration = 75 millis

  it should "tell the supervisor to create a new compute actor at startup" in {
    suppressStackTraceNoise {
      val resultProbe = TestProbe()
      val computeSupervisorProbe = TestProbe()
      createClientActor(computeSupervisorProbe.ref, resultProbe.ref, List())
      computeSupervisorProbe.expectMsgClass(Zero, classOf[CreateComputeActor])
    }
  }

  it should "receive a computeActor ref from the supervisor and delegate all risky work to it" in {
    suppressStackTraceNoise {
      val computeSupervisorProbe = TestProbe()
      val computeActorProbe = TestProbe()
      val work = List(RiskyAddition(1, 3), RiskyAddition(1, 5), RiskyAddition(6, 3))
      val clientActor = createClientActor(computeSupervisorProbe.ref, mock[ActorRef], work)

      computeSupervisorProbe.expectMsgClass(Zero, classOf[CreateComputeActor])
      clientActor.tell(computeActorProbe.ref, computeSupervisorProbe.ref)

      computeActorProbe.expectMsgClass(timeout, classOf[RiskyAddition])
      computeActorProbe.expectMsgClass(timeout, classOf[RiskyAddition])
      computeActorProbe.expectMsgClass(timeout, classOf[RiskyAddition])
    }
  }

  it should "stop if compute actor terminates" in {
    suppressStackTraceNoise {
      val computeTestActor = TestActorRef(Props(classOf[ComputeTestActor]))
      val computeSupervisorProbe = TestProbe()
      val clientActor = createClientActor(computeSupervisorProbe.ref, mock[ActorRef], List())

      computeSupervisorProbe.expectMsgClass(Zero, classOf[CreateComputeActor])
      computeSupervisorProbe.reply(computeTestActor)

      watch(clientActor)
      computeTestActor ! PoisonPill // Poison pill makes actor terminate
      expectMsgClass(timeout, classOf[Terminated])
    }
  }

  it should "compute and send risky work result to ResultActor when work throws no exceptions" in {
    suppressStackTraceNoise {
      val computeSupervisor = createComputeSupervisor
      val resultProbe = TestProbe()
      val work = List(RiskyAddition(2, 3), RiskyAddition(3, 3))
      createClientActor(computeSupervisor, resultProbe.ref, work)

      resultProbe.expectMsg(timeout, RiskyAdditionResult(5))
      resultProbe.expectMsg(timeout, RiskyAdditionResult(6))
    }
  }

  it should "compute and send remaining risky work to ResultActor when work throws risky work exception" in {
    suppressStackTraceNoise {
      class WorkWithFailure extends RiskyWork {
        override def perform() = throw new RiskyWorkException("test exception")
      }
      val computeSupervisor = createComputeSupervisor
      val resultProbe = TestProbe()
      val work = List(RiskyAddition(2, 3), new WorkWithFailure(), RiskyAddition(3, 3))
      createClientActor(computeSupervisor, resultProbe.ref, work)

      resultProbe.expectMsg(timeout, RiskyAdditionResult(5))
      resultProbe.expectMsg(timeout, RiskyAdditionResult(6))
    }
  }

  def createComputeSupervisor: TestActorRef[Nothing] = {
    val computeActorFactory = new ComputeActorFactory(mock[ActorRef])
    TestActorRef(Props(classOf[ComputeSupervisor], computeActorFactory))
  }

  def createClientActor(computeSupervisor: ActorRef, resultActor: ActorRef, riskyWork: List[RiskyWork]) = {
    TestActorRef(Props(classOf[ClientActor], computeSupervisor, resultActor, riskyWork))
  }
}

File: src/test/java/workshop/part2/ClientActorTest.java
package workshop.part2;

import akka.actor.*;
import akka.testkit.TestActorRef;
import akka.testkit.TestProbe;
import org.junit.Test;
import workshop.AkkaTest;
import workshop.work.RiskyWorkException;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.mockito.Mockito.mock;
import static workshop.AkkaTestHelper.getResult;
import static workshop.part2.ComputeSupervisor.CreateComputeActor;
import static workshop.work.Work.*;

public class ClientActorTest extends AkkaTest {

    @Test
    public void shouldTellTheSupervisorToCreateANewComputeActorAtStartup() {
        TestProbe resultProbe = TestProbe.apply(system);
        TestProbe computeSupervisorProbe = TestProbe.apply(system);

        List riskyWork = new ArrayList<RiskyWork>();
        createClientActor(computeSupervisorProbe.ref(), resultProbe.ref(), riskyWork);

        getResult(computeSupervisorProbe, CreateComputeActor.class);
    }

    @Test
    public void shouldReceiveComputeActorRefFromSupervisorAndDelegateAllRiskyWorkToIt() {
        List<RiskyWork> work = Arrays.asList(new RiskyAddition(1, 3), new RiskyAddition(1, 5), new RiskyAddition(6, 3));

        TestProbe computeSupervisorProbe = TestProbe.apply(system);
        TestProbe computeActorProbe = TestProbe.apply(system);
        TestActorRef<Actor> clientActor = createClientActor(computeSupervisorProbe.ref(), mock(ActorRef.class), work);

        getResult(computeSupervisorProbe, CreateComputeActor.class);
        clientActor.tell(computeActorProbe.ref(), computeSupervisorProbe.ref());

        getResult(computeActorProbe, RiskyAddition.class);
        getResult(computeActorProbe, RiskyAddition.class);
        getResult(computeActorProbe, RiskyAddition.class);
    }

    @Test
    public void shouldStopIfComputeActorTerminates() {
        TestActorRef<Actor> computeTestActor = TestActorRef.create(system, Props.create(ComputeTestActor.class));

        TestProbe computeSupervisorProbe = TestProbe.apply(system);
        TestActorRef<Actor> clientActor = createClientActor(computeSupervisorProbe.ref(), mock(ActorRef.class), new ArrayList<RiskyWork>());

        getResult(computeSupervisorProbe, CreateComputeActor.class);
        computeSupervisorProbe.reply(computeTestActor);

        TestProbe probe = TestProbe.apply(system);
        probe.watch(clientActor);

        computeTestActor.tell(PoisonPill.getInstance(), ActorRef.noSender()); // Poison pill makes actor terminate
        getResult(probe, Terminated.class);

    }

    @Test
    public void shouldComputeAndSendRiskyWorkResultToResultActorWhenWorkThrowsNoExceptions() {
        List<RiskyAddition> work = Arrays.asList(new RiskyAddition(2, 3), new RiskyAddition(3, 3));

        TestActorRef<Actor> computeSupervisor = createComputeSupervisor();
        TestProbe resultProbe = TestProbe.apply(system);
        createClientActor(computeSupervisor, resultProbe.ref(), work);

        assertEquals(5, getResult(resultProbe, RiskyAdditionResult.class).result);
        assertEquals(6, getResult(resultProbe, RiskyAdditionResult.class).result);
    }

    @Test
    public void shouldComputeAndSendRemainingRiskyWorkToResultActorWhenWorkThrowsRiskyWorkException() {
        List<RiskyWork> work = Arrays.asList(new RiskyAddition(2, 3), new WorkWithFailure(), new RiskyAddition(3, 3));

        TestActorRef<Actor> computeSupervisor = createComputeSupervisor();
        TestProbe resultProbe = TestProbe.apply(system);
        createClientActor(computeSupervisor, resultProbe.ref(), work);

        assertEquals(5, getResult(resultProbe, RiskyAdditionResult.class).result);
        assertEquals(6, getResult(resultProbe, RiskyAdditionResult.class).result);
    }

    public class WorkWithFailure extends RiskyWork {
        @Override
        public RiskyWorkResult perform() throws RiskyWorkException {
            throw new RiskyWorkException("test exception");
        }
    }

    private TestActorRef<Actor> createComputeSupervisor() {
        ComputeActorFactory computeActorFactory = new ComputeActorFactory(mock(ActorRef.class));
        return TestActorRef.create(system, Props.create(ComputeSupervisor.class, computeActorFactory));
    }

    private TestActorRef<Actor> createClientActor(ActorRef computeSupervisor, ActorRef resultActor, List riskyWork) {
        return TestActorRef.create(system, Props.create(ClientActor.class, computeSupervisor, resultActor, riskyWork));
    }
}